src/ir_Neoclima.cpp

```diff
@@ -194,19 +191,19 @@ void IRNeoclimaAc::setMode(const uint8_t mode) {
     case kNeoclimaCool:
     case kNeoclimaFan:
     case kNeoclimaHeat:
-      setBits(&remote_state[9], kNeoclimaModeOffset, kModeBitsSize, mode);
-      setButton(kNeoclimaButtonMode);
+      _.Mode = mode;
+      _.Button = kNeoclimaButtonMode;
       break;
     default:
       // If we get an unexpected mode, default to AUTO.
-      setMode(kNeoclimaAuto);
+      _.Mode = kNeoclimaAuto;
```

Comment:
This is not equivalent. It won't say the button value.
Reverting this line is better.

Reply:
you are right, it's wrong.
I want to eliminate recursive call, so it may look like this:
```
default:
    _.Mode = kNeoclimaAuto;
    _.Button = kNeoclimaButtonMode;
```
what do you think?

Reply:
That's fine. As long as the functionality is the same, I'm fine.

Longer term:
We should really look into if it is "cheaper" to do a recursive call, vs. duplicating the bit field setting.
i.e. Compare the number of instructions/bytes either generates to see which is more efficient flash-space-wise.
The main pressure we get from people is reducing the flash size, rather than the run-time stack size. i.e. Recursion may be a worthwhile alternative.

But that's for another time.

Reply:
I almost forgot that the original intention was to save flash space. It seems I shouldn't replace calls of getter/setter with duplicating the bit field setting. I need some time to test and compare.

Reply:
Well, it depends which saves us the most space ultimately. I don't know (yet).

Reply:
I use platformIO to build and compare the space size between V2.7.9 and V2.7.13 on wemos d1 mini,the configuration is as follows:

> CONFIGURATION: https://docs.platformio.org/page/boards/espressif8266/d1_mini.html
> PLATFORM: Espressif 8266 (2.6.2) > WeMos D1 R2 and mini
> HARDWARE: ESP8266 80MHz, 80KB RAM, 4MB Flash
> PACKAGES:
>  - framework-arduinoespressif8266 3.20704.0 (2.7.4)
>  - tool-esptool 1.413.0 (4.13)
>  - tool-esptoolpy 1.20800.0 (2.8.0)
>  - toolchain-xtensa 2.40802.200502 (4.8.2)
> Converting IRrecvDumpV2.ino
> LDF: Library Dependency Finder -> http://bit.ly/configure-pio-ldf
> LDF Modes: Finder ~ deep+, Compatibility ~ soft
> 
> Building in release mode

and the size of object file is as follows:

  | 2.7.13 | 2.7.9
-- | -- | --
ir_Airwell.cpp.o | 13,904 | 14,432
ir_Amcor.cpp.o | 17,544 | 17,968
ir_Argo.cpp.o | 23,920 | 24,636
ir_Carrier.cpp.o | 25,560 | 27,408
ir_Coolix.cpp.o | 35,948 | 35,176
ir_Corona.cpp.o | 26,784 | 27,764
ir_Daikin.cpp.o | 208,232 | 216,284
ir_Delonghi.cpp.o | 25,128 | 26,440
ir_Electra.cpp.o | 21,784 | 22,600
ir_Goodweather.cpp.o | 26,504 | 28,148
ir_Gree.cpp.o | 37,444 | 38,080
ir_Haier.cpp.o | 52,960 | 58,156
ir_Hitachi.cpp.o | 81,880 | 84,052
ir_Kelvinator.cpp.o | 24,732 | 26,612
ir_LG.cpp.o | 21,348 | 23,732
ir_Midea.cpp.o | 41,084 | 28,836
ir_Mitsubishi.cpp.o | 75,204 | 78,220
ir_MitsubishiHeavy.cpp.o | 62,912 | 64,908

As you can see, the size of object files is reduced except ir_Coolix.cpp.o and ir_Midea.cpp.o, which added new APIs.
And then I use platformIO inspect tools to analysis the memory use of examples, and the result is as follows:

  | 2.7.13 | 2.7.9
-- | -- | --
IRArgoAC | 26.5KB,260.1KB | 26.5KB,260.2KB
IRDaikinESP | 27.7KB,265.7KB | 27.6KB,266.0KB
IRGreeAC | 27.8KB,268.3KB | 27.8KB,268.3KB
IRKelvinatorAC | 27.8KB,265.3KB | 27.8KB,265.7KB
IRMitsubishiAC | 27.9KB,265.0KB | 27.8KB,265.3KB
IRMitsubishiHeavy152Ac | 27.8KB,265.1KB | 27.8KB,265.3KB

All of these example use less flash space except IRGreeAC, and IRDaikinESP and IRMitsubishiAC use about 100B more Ram while saving about 300B flash.

In summary, I think by reducing defined constants like k__offset, k__size, k__byte and using bit field dose save space.

And then I compare "do a recursive call, vs. duplicating the bit field setting".Let's call the original ir_Neoclima "V0",and after replacing getbits/setbits with bit field setting "V1", and then replace some calls of getter/setter with bit field setting "V2". The sizes of ir_Neoclima_0.o, ir_Neoclima_1.o, ir_Neoclima_2.o are respectively 50936B, 50272B, 49068B.
And then using
> xtensa-lx106-elf-size --format=SysV ir_Neoclima_0.o ir_Neoclima_1.o ir_Neoclima_2.o
the result is as follows after removing the same parts:

  | ir_Neoclima_0.o | ir_Neoclima_1.o | ir_Neoclima_2.o
-- | -- | -- | --
.literal | 148 | 212 | 332
.text | 5733 | 6081 | 6609
.rodata | 612 | 588 | 588
.gcc_except_table | 337 | 337 | 336
.eh_frame | 1452 | 1448 | 700
.xt.prop | 6504 | 6480 | 6480
Total | 15547 | 15907 | 15806

From thest three points ".literal",".text","Total", V0 uses the least space but it has the biggest object file size.

Then I use Arduino IDE and the example code

```
#include <Arduino.h>
#include <IRremoteESP8266.h>
#include <IRsend.h>
#include <ir_Neoclima.h>

const uint16_t kIrLed = 4;  // ESP8266 GPIO pin to use. Recommended: 4 (D2).
IRNeoclimaAc ac(kIrLed);  // Set the GPIO used for sending messages.

void printState() {
  // Display the settings.
  Serial.println("Neoclima A/C remote is in the following state:");
  Serial.printf("  %s\n", ac.toString().c_str());
  // Display the encoded IR sequence.
  unsigned char* ir_code = ac.getRaw();
  Serial.print("IR Code: 0x");
  for (uint8_t i = 0; i < kNeoclimaStateLength; i++)
    Serial.printf("%02X", ir_code[i]);
  Serial.println();
}

void setup() {
  ac.begin();
  Serial.begin(115200);
  delay(200);

  Serial.println("Default state of the remote.");
  printState();
  Serial.println("Setting desired state for A/C.");
  ac.setPower(true);  // Turn it on.
  ac.setFan(kNeoclimaFanMed);  // Medium Fan
  ac.setMode(kNeoclimaCool);  // Cool mode
  ac.setTemp(26);  // Celsius
  ac.setSwingV(true);  // Swing vertically
  ac.setSwingH(false);  // Swing Horizontally
}

void loop() {
  // Now send the IR signal.
  Serial.println("Sending IR command to A/C ...");
  ac.send();
  printState();
  delay(5000);
}
```

and compile it three times with three versions of ir_Neoclima, the result is:

> ir_Neoclima_0 :
> Executable segment sizes:
> IROM   : 238868          - code in flash         (default or ICACHE_FLASH_ATTR) 
> IRAM   : 27304   / 32768 - code in IRAM          (ICACHE_RAM_ATTR, ISRs...) 
> DATA   : 1248  )         - initialized variables (global, static) in RAM/HEAP 
> RODATA : 1828  ) / 81920 - constants             (global, static) in RAM/HEAP 
> BSS    : 25144 )         - zeroed variables      (global, static) in RAM/HEAP 
> 项目使用了 269248 字节，占用了 (25%) 程序存储空间。最大为 1044464 字节。
> 全局变量使用了28220字节，(34%)的动态内存，余留53700字节局部变量。最大为81920字节。
> 
> ir_Neoclima_1 :
> Executable segment sizes:
> IROM   : 238868          - code in flash         (default or ICACHE_FLASH_ATTR) 
> IRAM   : 27304   / 32768 - code in IRAM          (ICACHE_RAM_ATTR, ISRs...) 
> DATA   : 1248  )         - initialized variables (global, static) in RAM/HEAP 
> RODATA : 1828  ) / 81920 - constants             (global, static) in RAM/HEAP 
> BSS    : 25144 )         - zeroed variables      (global, static) in RAM/HEAP 
> 项目使用了 269248 字节，占用了 (25%) 程序存储空间。最大为 1044464 字节。
> 全局变量使用了28220字节，(34%)的动态内存，余留53700字节局部变量。最大为81920字节。
> 
> ir_Neoclima_2 :
> Executable segment sizes:
> IROM   : 238788          - code in flash         (default or ICACHE_FLASH_ATTR) 
> IRAM   : 27304   / 32768 - code in IRAM          (ICACHE_RAM_ATTR, ISRs...) 
> DATA   : 1248  )         - initialized variables (global, static) in RAM/HEAP 
> RODATA : 1828  ) / 81920 - constants             (global, static) in RAM/HEAP 
> BSS    : 25144 )         - zeroed variables      (global, static) in RAM/HEAP 
> 项目使用了 269168 字节，占用了 (25%) 程序存储空间。最大为 1044464 字节。
> 全局变量使用了28220字节，(34%)的动态内存，余留53700字节局部变量。最大为81920字节。

The V2 uses the least IROM. Finally, I decided to using the "duplicating the bit field setting" version.

These data are for information only, I hope there's a better way to analyze flash space.
Sorry, I'm not good at English.

Reply:
First off, that analysis is awesome! Top Marks!!

> Let's call the original ir_Neoclima "V0",and after replacing getbits/setbits with bit field setting "V1", and then replace some calls of getter/setter with bit field setting "V2".

I'm not sure what you mean between V1 & V2. i.e. What the difference is, but V2 seems to be the "winner" for small values of winner.

I'm in favour of that. So how should I program to achieve that. bit fields in setter/getters only? or everywhere?

Reply:
V1 replaces only getbits/setbits, V2 replaces both getbits/setbits and ***part*** of getter/setter.
I replace only recursive call and  getter/setter holds only one expression or can be replaced with one expression. For example:

```
uint8_t IRNeoclimaAc::getTemp(void) const {
  const uint8_t min_temp = getTempUnits() ? kNeoclimaMinTempC
                                          : kNeoclimaMinTempF;
  return _.Temp + min_temp;
}
......
uint8_t IRNeoclimaAc::getFan(void) const {
  return _.Fan;
}
```
I will replace "getFan()" with "_.Fan" but not replace "getTemp()";
For example:

```
void IRNeoclimaAc::setButton(const uint8_t button) {
  switch (button) {
    case kNeoclimaButtonPower:
    case kNeoclimaButtonMode:
    case kNeoclimaButtonTempUp:
    case kNeoclimaButtonTempDown:
    case kNeoclimaButtonSwing:
    case kNeoclimaButtonFanSpeed:
    case kNeoclimaButtonAirFlow:
    case kNeoclimaButtonHold:
    case kNeoclimaButtonSleep:
    case kNeoclimaButtonLight:
    case kNeoclimaButtonEye:
    case kNeoclimaButtonFollow:
    case kNeoclimaButtonIon:
    case kNeoclimaButtonFresh:
    case kNeoclimaButton8CHeat:
    case kNeoclimaButtonTurbo:
    case kNeoclimaButtonEcono:
    case kNeoclimaButtonTempUnit:
      _.Button = button;
      break;
    default:
      _.Button = kNeoclimaButtonPower;
  }
}
```
I will replace "setButton(kNeoclimaButtonPower)" with "_.Button = kNeoclimaButtonPower" within internal implemention.

Maybe someday we will test a version called V3 which replace getbits/setbits and ***every*** getter/setter :-)

Reply:
I think I get it now. Sounds good to me.
I see the fix now.